### DISCUSSION TITLE
Analogue to pip "editable installs"

### DISCUSSION THREAD

**Discussion by tomjakubowski** [pyodide/pyodide#4285] [Ideas]
for cpython, editable installs with `pip install -e` make it easy to test changes made in a python source package, without having to build + install a wheel:

```
pip install -e ~/path/to/mypackage
# imports from `mypackage` go through ~/path/to/mypackage, via a symlink in site-packages
python -c 'import mypackage; mypackage.foo()`
# change mypackage, edit foo
python -c 'import mypackage; mypackage.foo()`
# new foo() runs
```

Symlnks won't work for this style of pyodide development, at least when the Python code lives outside of pyodide's filesystem. If pyodide's importlib were able to support loading source packages from files served over http, maybe we could achieve a similar developer workflow in pyodide, something like this:

1. developer runs a file server, serving up their python package source tree (at, say, `https://files.localhost/path/to/mypackage`)
2. developer configures pyodide to use that file server to fetch contents of `mypackage` package:
```
loadPyodide({
  editable: {
    mypackage: 'https://files.localhost/path/to/mypackage'
  }
});
```
3. When importing `mypackage`, Pyodide importlib uses http to fetch `https://files.localhost/path/to/mypackage/__init__.py`, and then accesses the rest of the imports in `mypackage` via that same URL prefix.

This way, the developer workflow to test changes to a Python package in pyodide would be, pickign back up from step 2:

1. developer makes a change in `mypackage`
2. developer starts a fresh pyodide interpreter (by reloading the page, maybe)
3. new changes to `mypackage` are used this time

so the process would skip having to rebuild the wheel

Is this a viable idea? Would this feature be sufficiently useful above any alternatives (which I'd be unaware of, besides rebuilding the wheel) to be worth the cost? I am pretty ignorant of Python's import and package loading mechanisms (i.e. I know just enough to reference "importlib" :-))

**Comment by ryanking13** [comment]
Thanks for opening the discussion. There was a similar idea that I proposed and discarded: https://github.com/pyodide/pyodide/issues/1917, https://github.com/ryanking13/pyodide-importer.

There were several reasons that the idea was discarded:

1. It is almost impossible to find "the rest of the imports" correctly. As python import and packaging system is too dynamic.
2. If the package contains any compiled modules, it is hard to deal with, as we need to compile the module asynchronously.
3. If your package has a large number of files, you're going to get a ton of requests.

So, in my opinion, the probability of a similar feature being added to Pyodide is very low.

However, it is possible to hack importlib to add that functionality on your own (without adding it to Pyodide directly). Looking at https://github.com/ryanking13/pyodide-importer may be of some help to you.
